### PR TITLE
Check if the marshalling of a field is possible across all platforms.

### LINKED ISSUE #53110: AF: (SIZE_T)mtype < COUNTOF(nativeSizes) when building COM vtable
https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-52992-merge-1da0184fed3444a4bd/System.Runtime.Tests/console.00381496.log?sv=2019-07-07&se=2021-06-10T21%3A55%3A13Z&sr=c&sp=rl&sig=Na%2FXK1cDDkv8h87piXD8wapDbv4gAyoGP0th40V9x6k%3D

https://dev.azure.com/dnceng/public/_build/results?buildId=1150299&view=logs&j=bbd47f15-9f4c-590a-7d78-32b645974ead&t=5f0d06d6-9390-514d-fa1d-79fe5d3023fe&l=56

```
Console log: 'System.Runtime.Tests' from job 1da0184f-ed34-44a4-bd6d-fa09fa6457c9 workitem f3c909b1-80b0-48f4-b9ac-df8309a98c73 (windows.10.amd64.open.rt) executed on machine a0049N5

C:\h\w\B46309D4\w\AF600982\e>call RunTests.cmd --runtime-path C:\h\w\B46309D4\p 
----- start Fri 05/21/2021 22:18:30.18 ===============  To repro directly: ===================================================== 
pushd C:\h\w\B46309D4\w\AF600982\e\
"C:\h\w\B46309D4\p\dotnet.exe" exec --runtimeconfig System.Runtime.Tests.runtimeconfig.json --depsfile System.Runtime.Tests.deps.json xunit.console.dll System.Runtime.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing 
popd
===========================================================================================================

C:\h\w\B46309D4\w\AF600982\e>"C:\h\w\B46309D4\p\dotnet.exe" exec --runtimeconfig System.Runtime.Tests.runtimeconfig.json --depsfile System.Runtime.Tests.deps.json xunit.console.dll System.Runtime.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing  
  Discovering: System.Runtime.Tests (method display = ClassAndMethod, method display options = None)

Assert failure(PID 4564 [0x000011d4], Thread: 9668 [0x25c4]): (SIZE_T)mtype < COUNTOF(nativeSizes)

CORECLR! GetCLRRuntimeHost + 0x2A92BC (0x71fbeb0c)
CORECLR! GetCLRRuntimeHost + 0x2AD03C (0x71fc288c)
CORECLR! GetCLRRuntimeHost + 0x3CF71B (0x720e4f6b)
CORECLR! GetCLRRuntimeHost + 0x3CF175 (0x720e49c5)
CORECLR! GetCLRRuntimeHost + 0x3B86EC (0x720cdf3c)
CORECLR! GetCLRRuntimeHost + 0x3B0678 (0x720c5ec8)
CORECLR! GetCLRRuntimeHost + 0x3AEDEE (0x720c463e)
CORECLR! GetCLRRuntimeHost + 0x3AE561 (0x720c3db1)
CORECLR! GetCLRRuntimeHost + 0x231669 (0x71f46eb9)
CORECLR! GetCLRRuntimeHost + 0x2435D2 (0x71f58e22)
    File: D:\workspace\_work\1\s\src\coreclr\vm\mlinfo.cpp Line: 2930
    Image: C:\h\w\B46309D4\p\dotnet.exe
```

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Fixes #53110 

The x86 code path for field marshalling required inspection of the field earlier than the amd64 code path. This update makes that inspection occur at the same time in both cases. However, the only real change here is that both will fail later on if the field is attempted to be marshalled via the IClassX interface. The entire experience here is poor though.

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
@jkoritzinsky and @elinor-fung This is a draft PR to see what people think of the fix. It basically makes the x86 and amd64 logic consistent but it is still poor. We also need a test here.

**Comment by jkoritzinsky** [comment]
I'd be ok doing the validation like this.

Personally, I wish we didn't need to support IClassX-style interface generation at all (especially since tlbexp isn't supported), but there's not much we can do about that.

**Comment by AaronRobinsonMSFT** [comment]
Failures are https://github.com/dotnet/runtime/issues/53350 https://github.com/dotnet/runtime/issues/53351 and https://github.com/dotnet/runtime/issues/53353.

**Review by jkoritzinsky — APPROVED** [review]

File: src/coreclr/vm/comtoclrcall.cpp [hunk 1]
```diff
@@ -1030,6 +1029,10 @@ void ComCallMethodDesc::InitNativeInfo()
 #endif
                              );
 
+            if (info.GetMarshalType() == MarshalInfo::MARSHAL_TYPE_UNKNOWN)
+                info.ThrowTypeLoadExceptionForInvalidFieldMarshal(pFD, info.GetErrorResourceId());
```
Comment (elinor-fung) [inline src/coreclr/vm/comtoclrcall.cpp:1033]: Do we have a test for the failure case?

**Pull request merged**